### PR TITLE
Nav Unification: Add Emails submenu to the Upgrades navigation menu

### LINKED ISSUE #51150: Nav Unification: "domains" sub menu position not working
In the old Calypso navigation scheme, the Upgrades menu is:

- Upgrades
 - Plans
 - Domains
 - Purchases (where you manage both plan and domains purchases)

However, in the unified navigation it's:

- Upgrades
 - Plans
 - Purchases (where you manage both plan and domains purchases)
 - Domains

In `WPcom_Admin_Menu` the [position argument for the Domains sub menu](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/modules/masterbar/admin-menu/class-wpcom-admin-menu.php#L201) seems to be correct (10, where Plans is 5 and Purchases is 15), but isn't respected.

Long story short, it makes sense to have purchase management (Purchases) at the bottom of this sub-menu, especially as we add other items to the menu (i.e. email).

Reported in: p9Jlb4-2iV-p2

### PR DESCRIPTION
**Pull request by hambai**
We'll need to add Emails under Upgrades once the new Emails home page is ready -- we're expecting this for . While I was at it I noticed that the ordering of the submenus is off so here's a fix for that too.

**Note:** Please hold off on merging this as we're working on improving the WPCOM UI. As a side note, what is the best way to restrict access to Automattic staff? We'd like to enable interim testing before we launch the new menu item to users.

Fixes https://github.com/Automattic/wp-calypso/issues/51150

#### Changes proposed in this Pull Request:

* Add the Emails submenu under Upgrades in the Simple and Atomic navigation
* Fix the Upgrades submenus order

#### Testing instructions:

**WordPress.com Simple**
* Merge the changed files from D59693-code on your sandbox
* Without making any changes on your sandbox, verify that the navigation menu under Upgrades is in the following order: `Plans, Domains, Purchases`
* Now add the following code to run on your sandbox:
```
add_filter( 'jetpack_show_wpcom_upgrades_email_menu', '__return_true' );
```
* Verify that the navigation menu under Upgrades is in the following order: `Plans, Domains, Emails, Purchases`

**WordPress.com Atomic**

* Ensure that you have the [[ https://jetpack.com/download-jetpack-beta/ | Jetpack Beta ]] plugin uploaded and installed on your Atomic test site, as well as a mechanism for running code on the site. (A plugin like [[ https://wordpress.org/plugins/code-snippets/ | Code Snippets ]] can make this relatively easy.)
* Use Jetpack Beta to switch to this branch: `add/emails-submenu-to-upgrades`
* Refresh the admin page you're on, and verify that the Upgrades menu contains the following submenus in this sequence: `Plans, Domains, Purchases`
* Navigate to "My Home" (which is served via Calypso), and verify that the Upgrades menu shows the `Plans, Domains, Purchases` items in that sequence. (It may take a little bit to update, and you may need to clear the Calypso state to make sure the menu gets updated.)
* Ensure that the following code is run on your site:
```
add_filter( 'jetpack_show_wpcom_upgrades_email_menu', '__return_true' );
```
* Load a new admin page, and verify that the Upgrades menu includes the following submenu items in this order: `Plans, Domains, Emails, Purchases`
* Navigate to "My Home" and verify that the Upgrades menu now includes `Plans, Domains, Emails, Purchases`. As above, you may need to clear the Calypso data and/or wait for the REST response to get applied.

**Here's a screenshot:**
Screenshot 2021-04-23 at 11 42 47

#### Does this pull request change what data or activity we track or use?

No, this PR does not have any privacy impact -- it simply adds a new submenu item and fixes an ordering bug in the submenu.

### CODE REVIEW

**Review by olaseni — DISMISSED** [review]
LGTM! The new menu is added, and the additional item fix looks great.

**Comment by jeherve** [comment]
Would this now be ready to review, and possibly merge? The "Do not merge" in the title is making me doubt :)

**Comment by jeherve** [comment]
> As a side note, what is the best way to restrict access to Automattic staff? We'd like to enable interim testing before we launch the new menu item to users.

That's not something we typically do in the Jetpack plugin. You could, however use a filter to gate the change, and ask automatticians to set the filter to `true` to access the change, for example, or even set that filter to true via wpcomsh for automatticians, if necessary.

cc @mmtr who has some experience with nav-unification gradual changes like this.

**Comment by daledupreez** [comment]
Thanks for the prompt review, @jeherve!

From an "Is the PR working as expected?" perspective, this is ready for review. The "DO NOT MERGE" label is there to make sure we don't merge before the related features in Calypso are ready -- they are still behind a feature flag and won't be ready for launch until the middle of next week at least. We got started on this PR ahead of time because we weren't sure how long it would take us to get any changes deployed!

I ended up borrowing from the approach in #18105 using a filter that defaults to `false`, but I am 100% open to renaming the filter. If the filter looks OK to you all, then I think we can remove the "DO NOT MERGE" protection for now - we can then roll out a smaller PR when needed to remove the check once we're ready to launch.

**Comment by daledupreez** [comment]
This isn't working as expected on WPCOM -- I'll dig in more tomorrow.

**Comment by jeherve** [comment]
> The "DO NOT MERGE" label is there to make sure we don't merge before the related features in Calypso are ready -- they are still behind a feature flag and won't be ready for launch until the middle of next week at least. We got started on this PR ahead of time because we weren't sure how long it would take us to get any changes deployed!

Thanks for the extra details! To clarify (or add a bit of complexity :) ) to the situation:

A new version of Jetpack will be released on May 4. On that day, all Atomic sites will receive any changes that were included in the codebase this month. Since we want to test everything before the release, we'll have a code freeze from Monday, EOD. Our Beta testers will then have a week to test everything. 
The next release will then be on June 1.

In practice, that means that if you'd like to enable something for Atomic site owners sometime in May, you'll need to have this PR merged before Monday EOD. I would suggest keeping this date as a target. Even if the Calypso PR isn't merged yet, you can get this Jetpack PR merged with the filter set in wpcomsh, allowing you to effectively flip the switch later, after May 4 and once the Calypso PR is merged.

Does that make sense?

Don't hesitate to reach out in #jetpack-crew if you have any questions!

File: projects/plugins/jetpack/modules/masterbar/admin-menu/class-atomic-admin-menu.php [hunk 1]
```diff
@@ -226,7 +226,12 @@ public function add_upgrades_menu( $plan = null ) {
 		}
 		parent::add_upgrades_menu( $plan );
 
-		add_submenu_page( 'paid-upgrades.php', __( 'Domains', 'jetpack' ), __( 'Domains', 'jetpack' ), 'manage_options', 'https://wordpress.com/domains/manage/' . $this->domain, null, 10 );
+		$last_upgrade_submenu_position = $this->get_submenu_item_count( 'paid-upgrades.php' );
+
+		add_submenu_page( 'paid-upgrades.php', __( 'Domains', 'jetpack' ), __( 'Domains', 'jetpack' ), 'manage_options', 'https://wordpress.com/domains/manage/' . $this->domain, null, $last_upgrade_submenu_position - 1 );
+		if ( apply_filters( 'jetpack_show_wpcom_upgrades_email_menu', false ) ) {
```
Comment (jeherve) [inline projects/plugins/jetpack/modules/masterbar/admin-menu/class-atomic-admin-menu.php:232]: You'll want to add a docblock to that filter, so the Jetpack Codex can pick up the information when the hooks are parsed to create documentation. 

Here is an example:
https://github.com/Automattic/jetpack/blob/d3f3117217ca825f63eebbf7784ad9ef7140bec5/projects/plugins/jetpack/modules/masterbar.php#L30-L39

**Review by jeherve — CHANGES_REQUESTED** [review]

File: projects/plugins/jetpack/modules/masterbar/admin-menu/class-atomic-admin-menu.php (hunk 1 above)
Comment (daledupreez) [reply projects/plugins/jetpack/modules/masterbar/admin-menu/class-atomic-admin-menu.php:243]: Done!

File: projects/plugins/jetpack/modules/masterbar/admin-menu/class-wpcom-admin-menu.php [hunk 2]
```diff
@@ -203,7 +203,23 @@ public function add_upgrades_menu( $plan = null ) {
...
+		/**
+		 * Whether to show the WordPress.com Emails submenu under the main Upgrades menu.
+		 *
+		 * @use add_filter( 'jetpack_show_wpcom_upgrades_email_menu', '__return_true' );
+		 * @module masterbar
+		 *
+		 * @since 9.7.0
+		 *
+		 * @param bool $show_wpcom_upgrades_email_menu Load the WordPress.com Emails submenu item. Default to false.
+		 */
```
Comment (jeherve) [inline projects/plugins/jetpack/modules/masterbar/admin-menu/class-wpcom-admin-menu.php:219]:
```suggestion
		/** This filter is already documented in modules/masterbar/admin-menu/class-atomic-admin-menu.php */
```

File: projects/plugins/jetpack/tests/php/modules/masterbar/test-class-atomic-admin-menu.php [hunk 3]
```diff
@@ -272,7 +272,14 @@ public function test_add_upgrades_menu() {
 
 		static::$admin_menu->add_upgrades_menu();
 
-		$this->assertSame( 'https://wordpress.com/domains/manage/' . static::$domain, array_pop( $submenu['paid-upgrades.php'] )[2] );
+		$this->assertSame( 'https://wordpress.com/plans/my-plan/' . static::$domain, $submenu['paid-upgrades.php'][1][2] );
+		$this->assertSame( 'https://wordpress.com/domains/manage/' . static::$domain, $submenu['paid-upgrades.php'][2][2] );
+		if ( apply_filters( 'jetpack_show_wpcom_upgrades_email_menu', false ) ) {
```
Comment (jeherve) [inline projects/plugins/jetpack/tests/php/modules/masterbar/test-class-atomic-admin-menu.php:277]:
```suggestion
		/** This filter is already documented in modules/masterbar/admin-menu/class-atomic-admin-menu.php */
		if ( apply_filters( 'jetpack_show_wpcom_upgrades_email_menu', false ) ) {
```

File: projects/plugins/jetpack/tests/php/modules/masterbar/test-class-wpcom-admin-menu.php [hunk 4]
```diff
@@ -251,7 +251,14 @@ public function test_add_upgrades_menu() {
 
 		static::$admin_menu->add_upgrades_menu();
 
-		$this->assertSame( 'https://wordpress.com/domains/manage/' . static::$domain, array_pop( $submenu['paid-upgrades.php'] )[2] );
+		$this->assertSame( 'https://wordpress.com/plans/my-plan/' . static::$domain, $submenu['paid-upgrades.php'][1][2] );
+		$this->assertSame( 'https://wordpress.com/domains/manage/' . static::$domain, $submenu['paid-upgrades.php'][2][2] );
+		if ( apply_filters( 'jetpack_show_wpcom_upgrades_email_menu', false ) ) {
```
Comment (jeherve) [inline projects/plugins/jetpack/tests/php/modules/masterbar/test-class-wpcom-admin-menu.php:256]:
```suggestion
		/** This filter is already documented in modules/masterbar/admin-menu/class-atomic-admin-menu.php */
		if ( apply_filters( 'jetpack_show_wpcom_upgrades_email_menu', false ) ) {
```

**Review by jeherve — CHANGES_REQUESTED** [review]

File: projects/plugins/jetpack/tests/php/modules/masterbar/test-class-wpcom-admin-menu.php (hunk 4 above)
Comment (daledupreez) [reply projects/plugins/jetpack/tests/php/modules/masterbar/test-class-wpcom-admin-menu.php:258]: I've fixed/improved the filter docs in e64f25c5d3e8dbca7c5b8fe0bf3e7ec1b98d8e22.
Apologies for the duplication!

Comment (jeherve) [reply projects/plugins/jetpack/tests/php/modules/masterbar/test-class-wpcom-admin-menu.php:258]: No worries! That's a bit of a specificity of the Jetpack codex. :)

**Review by jeherve — APPROVED** [review]

**Comment by daledupreez** [comment]
Thanks for the ultra-quick reviews, @jeherve!

Is there any reason for us to hold off merging this until Monday? I know I am not going to deploy the pieces for WordPress.com until Monday anyway, so I don't want to merge something that may cause trouble over the weekend.

**Comment by jeherve** [comment]
> Is there any reason for us to hold off merging this until Monday? I know I am not going to deploy the pieces for WordPress.com until Monday anyway, so I don't want to merge something that may cause trouble over the weekend.

I would suggest merging it right away, to avoid having to rebase on Monday if someone else works on that code until then. In any case, nothing will happen once it's merged; it won't be used on any production site, and dev sites won't see a difference unless they use the filter.

**Comment by daledupreez** [comment]
Thanks for confirming, @jeherve! 🚀

**Comment by daledupreez** [comment]
This change was committed in r224707-wpcom.

**Pull request merged**